cloud-controller-manager/do/droplets.go

```diff
@@ -159,51 +75,52 @@ func (i *instances) InstanceExistsByProviderID(ctx context.Context, providerID s
-	dropletID, err := dropletIDFromProviderID(providerID)
+func (i *instances) InstanceShutdown(ctx context.Context, node *v1.Node) (bool, error) {
+	dropletID, err := dropletIDFromProviderID(node.Spec.ProviderID)
 	if err != nil {
-		return false, fmt.Errorf("error getting droplet ID from provider ID %q: %s", providerID, err)
+		return false, fmt.Errorf("determining droplet ID from providerID: %s", err.Error())
 	}
 
 	droplet, err := dropletByID(ctx, i.resources.gclient, dropletID)
 	if err != nil {
-		return false, fmt.Errorf("error getting droplet \"%d\" by ID: %s", dropletID, err)
+		return false, fmt.Errorf("getting droplet by ID: %s: ", err.Error())
+	}
+	if droplet == nil {
+		return false, fmt.Errorf("droplet %d for node %s does not exist", dropletID, node.Name)
 	}
 
 	return droplet.Status == dropletShutdownStatus, nil
 }
 
-// dropletByID returns a *godo.Droplet value for the droplet identified by id.
-func dropletByID(ctx context.Context, client *godo.Client, id int) (*godo.Droplet, error) {
-	droplet, _, err := client.Droplets.Get(ctx, id)
-	return droplet, err
-}
-
-// dropletByName returns a *godo.Droplet for the droplet identified by nodeName.
-//
-// When nodeName identifies more than one droplet, only the first will be
-// considered.
-func dropletByName(ctx context.Context, client *godo.Client, nodeName types.NodeName) (*godo.Droplet, error) {
-	// TODO (andrewsykim): list by tag once a tagging format is determined
-	droplets, err := allDropletList(ctx, client)
+func (i *instances) InstanceMetadata(ctx context.Context, node *v1.Node) (*cloudprovider.InstanceMetadata, error) {
+	dropletID, err := dropletIDFromProviderID(node.Spec.ProviderID)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("determining droplet ID from providerID: %s", err.Error())
 	}
 
-	for _, droplet := range droplets {
-		if droplet.Name == string(nodeName) {
-			return &droplet, nil
-		}
-		addresses, _ := nodeAddresses(&droplet)
-		for _, address := range addresses {
-			if address.Address == string(nodeName) {
-				return &droplet, nil
-			}
-		}
+	droplet, err := dropletByID(ctx, i.resources.gclient, dropletID)
+	if err != nil {
+		return nil, fmt.Errorf("getting droplet by ID: %s: ", err.Error())
+	}
+	if droplet == nil {
+		return nil, fmt.Errorf("droplet %d for node %s does not exist", dropletID, node.Name)
 	}
+	nodeAddrs, err := nodeAddresses(droplet)
+	if err != nil {
+		return nil, fmt.Errorf("getting node addresses of droplet %d for node %s: %s", dropletID, node.Name, err.Error())
+	}
+	return &cloudprovider.InstanceMetadata{
+		ProviderID:    node.Spec.ProviderID, // the providerID may or may not be present according to the interface doc. However, we set this from kubelet.
+		InstanceType:  droplet.SizeSlug,
+		Region:        droplet.Region.Slug,
+		NodeAddresses: nodeAddrs,
+	}, nil
+}
 
-	return nil, cloudprovider.InstanceNotFound
```

Comment:
Ah, I need to handle this in InstancesV2

Reply:
Handled, with a note that returning false with no error from InstanceExists will remove the node from the cluster.  We return nil error if godo err response is http.StatusNotFound.